File: adapters/python/tests/fake_writ.py
"""A tiny Contract 6 gateway for tests: ``fake_writ.py [--policy P] [--ledger L] check --stdio --ask deny|defer``.

The verdict is chosen by tool name (no real policy):

    allowed / anything else  -> allow
    denied / bash + "rm -rf" -> deny (rule no-rm, writ.yaml:12)
    asky                     -> ask (approval required under --ask defer)
    redacty                  -> redact, pattern \\d{3}-\\d{2}-\\d{4}
    malformed                -> writes a non-JSON line
    hang                     -> never answers
    crash                    -> exits with code 3
    error                    -> {"error": {...}}
    wrongid                  -> answers with a different id
    liar                     -> deny with dispatch=true (contradiction)
    redact_noout             -> redact, but complete omits output

Every request is appended as a JSON line to ``$FAKE_WRIT_LOG`` when set.
Set ``FAKE_WRIT_FAIL=1`` to exit 1 immediately, like the stub binary.
"""

from __future__ import annotations

import json
import os
import re
import sys
import time

SSN = r"\d{3}-\d{2}-\d{4}"


def main() -> int:
    argv = sys.argv[1:]
    if os.environ.get("FAKE_WRIT_FAIL") == "1":
        sys.stderr.write("writ check is not implemented in this build (fail closed)\n")
        return 1
    ask_mode = "deny"
    if "--ask" in argv:
        ask_mode = argv[argv.index("--ask") + 1]
    if "check" not in argv or "--stdio" not in argv:
        sys.stderr.write(f"fake_writ: unexpected argv {argv}\n")
        return 1
    log_path = os.environ.get("FAKE_WRIT_LOG")
    refs: dict[str, dict] = {}
    n = 0
    out = sys.stdout

    def send(obj: dict) -> None:
        out.write(json.dumps(obj) + "\n")
        out.flush()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        req = json.loads(line)
        if log_path:
            with open(log_path, "a", encoding="utf-8") as f:
                f.write(json.dumps({"argv": argv, **req}) + "\n")
        rid = req.get("id")
        base = {"v": 1, "id": rid}
        op = req.get("op")
        if op == "decide":
            call = req["call"]
            tool = call.get("tool")
            n += 1
            ref = f"ref-{n}"
            refs[ref] = {"tool": tool, "call": call}
            cmd = (call.get("args") or {}).get("command", "")
            if tool == "malformed":
                out.write("this is not json\n")
                out.flush()
                continue
            if tool == "hang":
                time.sleep(3600)
            if tool == "crash":
                return 3
            if tool == "error":
                send({**base, "error": {"code": "bad_request", "message": "fake failure"}})
                continue
            if tool == "wrongid":
                send({**base, "id": "nope", "decision": "allow", "dispatch": True, "ref": ref})
                continue
            if tool == "liar":
                send({**base, "decision": "deny", "dispatch": True, "rule_id": "x", "ref": ref})
                continue
            if tool == "denied" or (tool == "bash" and "rm -rf" in cmd):
                send({**base, "decision": "deny", "dispatch": False, "rule_id": "no-rm",
                      "reason": "Destructive command.", "location": "writ.yaml:12", "ref": ref})
                continue
            if tool == "asky":
                if ask_mode == "defer":
                    send({**base, "decision": "ask", "dispatch": False, "approval": "required",
                          "rule_id": "prod", "reason": "--- diff ---", "irreversible": True,
                          "timeout_ms": 5000, "ref": ref})
                else:  # what the real gateway sends for an ask under --ask deny
                    send({**base, "decision": "deny", "verdict": "ask", "dispatch": False,
                          "rule_id": "prod", "reason": "requires human approval (fail closed)",
                          "location": "writ.yaml:20", "ref": ref})
                continue
            if tool in ("redacty", "redact_noout"):
                send({**base, "decision": "redact", "dispatch": True, "rule_id": "pii",
                      "patterns": [SSN], "ref": ref})
                continue
            send({**base, "decision": "allow", "dispatch": True, "rule_id": "read-only", "ref": ref})
        elif op == "resolve":
            ref = req.get("ref")
            if ref not in refs:
                send({**base, "error": {"code": "bad_ref", "message": "unknown ref"}})
                continue
            if req.get("approved") is True:
                new_ref = f"{ref}.approved"  # like writ: an approval mints a new ref
                refs[new_ref] = {**refs[ref], "approved": True}
                send({**base, "decision": "allow", "dispatch": True, "rule_id": "prod", "ref": new_ref})
            else:
                send({**base, "decision": "deny", "dispatch": False, "rule_id": "prod",
                      "reason": "rejected by approver", "ref": ref})
        elif op == "complete":
            ref = req.get("ref")
            if ref not in refs:
                send({**base, "error": {"code": "bad_ref", "message": "unknown ref"}})
                continue
            tool = refs[ref]["tool"]
            if tool == "asky" and not refs[ref].get("approved"):
                send({**base, "error": {"code": "state", "message": "complete for an ask that was not approved"}})
                continue
            resp = {**base, "recorded": True}
            if tool == "redacty":
                resp["output"] = re.sub(SSN, "[redacted-by-writ]", req.get("output") or "")
            send(resp)
        else:
            send({**base, "error": {"code": "bad_request", "message": f"unknown op {op!r}"}})
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: adapters/typescript/test/fixtures/fake-writ.mjs
#!/usr/bin/env node
// A fake `writ check --stdio` implementing INTERFACES.md Contract 6 (v1) for
// tests. Behaviour is chosen by the call's `tool`:
//   deny | ask | redact | malformed | hang | crash | error | wrongid | noref |
//   contradict | crash-on-complete | anything else => allow.
// FAKE_WRIT_LOG=<file> appends {argv} and every request as JSON lines.
// FAKE_WRIT_STARTUP_CRASH=1 exits immediately.
import { appendFileSync } from "node:fs";

const argv = process.argv.slice(2);
const log = (obj) => {
  if (process.env.FAKE_WRIT_LOG) appendFileSync(process.env.FAKE_WRIT_LOG, JSON.stringify(obj) + "\n");
};
log({ argv });

if (process.env.FAKE_WRIT_STARTUP_CRASH === "1") {
  process.stderr.write("fake writ: startup failure\n");
  process.exit(1);
}
const checkAt = argv.indexOf("check");
if (checkAt < 0 || !argv.includes("--stdio")) {
  process.stderr.write("fake writ: expected `check --stdio`\n");
  process.exit(1);
}
const askAt = argv.indexOf("--ask");
const askMode = askAt >= 0 ? argv[askAt + 1] : "deny";

const refs = new Map(); // ref -> { tool, decision, patterns, state }
let n = 0;
const send = (obj) => process.stdout.write(JSON.stringify({ v: 1, ...obj }) + "\n");

function decide(id, call) {
  const tool = call?.tool;
  const ref = `ref-${++n}`;
  switch (tool) {
    case "deny":
      refs.set(ref, { tool, state: "denied" });
      return send({ id, decision: "deny", dispatch: false, rule_id: "no-rm", reason: "Destructive command.", location: "writ.yaml:7", ref });
    case "ask": {
      const timeout_ms = typeof call.args?.timeout_ms === "number" ? call.args.timeout_ms : 60000;
      if (askMode === "defer") {
        refs.set(ref, { tool, state: "deferred" });
        return send({ id, decision: "ask", dispatch: false, approval: "required", rule_id: "prod", reason: "diff: DROP TABLE users", irreversible: true, timeout_ms, ref });
      }
      // As the real gateway: --ask deny answers deny, tagged with the original verdict.
      refs.set(ref, { tool, state: "denied" });
      return send({ id, decision: "deny", verdict: "ask", dispatch: false, rule_id: "prod", reason: "requires human approval; --ask deny has no approver", location: "writ.yaml:12", ref });
    }
    case "redact":
      refs.set(ref, { tool, state: "dispatched", redact: true });
      return send({ id, decision: "redact", dispatch: true, rule_id: "pii", patterns: ["[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}"], ref });
    case "malformed":
      return process.stdout.write("this is not json\n");
    case "hang":
      return;
    case "crash":
      process.stderr.write("fake writ: boom\n");
      return process.exit(3);
    case "error":
      return send({ id, error: { code: "policy_error", message: "policy failed to load" } });
    case "wrongid":
      return send({ id: "not-" + id, decision: "allow", dispatch: true, ref });
    case "noref":
      return send({ id, decision: "allow", dispatch: true, rule_id: "x" });
    case "contradict":
      return send({ id, decision: "deny", dispatch: true, rule_id: "x", ref });
    default:
      refs.set(ref, { tool, state: "dispatched", crashOnComplete: tool === "crash-on-complete" });
      return send({ id, decision: "allow", dispatch: true, rule_id: "allow-all", ref });
  }
}

function handle(req) {
  const { id, op } = req;
  if (req.v !== 1) return send({ id, error: { code: "bad_request", message: "unsupported protocol version" } });
  if (op === "decide") return decide(id, req.call);
  const entry = refs.get(req.ref);
  if (op === "resolve") {
    if (!entry || entry.state !== "deferred") return send({ id, error: { code: "bad_request", message: "resolve without a deferred ask" } });
    if (req.approved === true) {
      // As the real gateway: an approval yields a new ref; only that one completes.
      entry.state = "resolved";
      const approvedRef = `${req.ref}-approved`;
      refs.set(approvedRef, { tool: entry.tool, state: "dispatched" });
      return send({ id, decision: "allow", dispatch: true, rule_id: "prod", ref: approvedRef });
    }
    entry.state = "denied";
    return send({ id, decision: "deny", dispatch: false, rule_id: "prod", reason: "approval rejected", ref: req.ref });
  }
  if (op === "complete") {
    if (!entry || entry.state !== "dispatched") return send({ id, error: { code: "bad_request", message: "complete without a dispatched decision" } });
    if (entry.crashOnComplete) return process.exit(4);
    entry.state = "completed";
    if (entry.redact && typeof req.output === "string") {
      return send({ id, recorded: true, output: req.output.replace(/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g, "[REDACTED]") });
    }
    return send({ id, recorded: true });
  }
  return send({ id, error: { code: "bad_request", message: `unknown op ${String(op)}` } });
}

let buf = "";
process.stdin.setEncoding("utf8");
process.stdin.on("data", (chunk) => {
  buf += chunk;
  let i;
  while ((i = buf.indexOf("\n")) >= 0) {
    const line = buf.slice(0, i);
    buf = buf.slice(i + 1);
    if (!line.trim()) continue;
    let req;
    try {
      req = JSON.parse(line);
    } catch {
      send({ id: null, error: { code: "bad_request", message: "malformed request" } });
      continue;
    }
    log(req);
    handle(req);
  }
});
process.stdin.on("end", () => process.exit(0));
